### ISSUE TITLE
[Proposal] - Better added value for Vault (or OpenBao)

### ISSUE THREAD

**Issue by gillg** [notaryproject/notation-hashicorp-vault#28]
# Current situation

Hashicorp Vault, or its fork OpenBao are designed to securely store self-hosted secrets and provide an auditable access to them.
One of its power resides in the multiple kind of secrets it can host via engines. The base is obviosly a simple key value database, and it's what this current notation plugin supports. But plenty of other tools are doing that, cloud managed, local, self managed, because encrypting key/values strings is the base, but not really a challenge. Using Vault just for that is probably a missunderstanding of what can be done.

Storing a **private** certificate signature key on it, means you generated it localy, and nothing proves that key were not shared before in an internal wiki/git/whatever, and any admin of vault (at the minimum) can exfiltate the key in the future. You have no way to know if it has been used outside of your supply chain, once you discover it it's too late...

# Proposed improvement

As we consider users of this plugin already have a Vault (or are able to manage a vault), the PKI secrets engine fits a lot more for a secure code signing.
Using PKI engine instead of key/value, you can generate a CA cert and private key directly on vault (not locally with openSSL) without never having access to the private key.
Then you call the /sign API to derivate sub signing keys with a short life time. That means we get the private key locally (like with kv) but it can't be stollen and reused later thanks to the short life. In the trust policy of Notary, you trust your root certificate from Vault and its derivate "short life" signature certificates.
Limitation: you can't use a cert provided by a public root CA as you can't import the private key or make a CSR.

Example https://www.hashicorp.com/blog/code-signing-with-hashicorp-vault-and-github-actions

# Alternatives / future steps

If you use Hachicorp vault enterprise, you can even ask for EV CodeSigning certs to a public CA because your vault can be FIPS 140-2 level 2 (required for a public EV cert).
And **maybe** thanks to the enterprise feature EST https://developer.hashicorp.com/vault/docs/secrets/pki/est it could be possible to generate a leaf cert (instead of a CA), and get the CSR from it then re-import the final issued cert inside vault.
This is what is possible also via "Azure KeyVault", but the main benefit of Vault is the cloud agnostic and self-hosted approach.

**Comment by cipherboy** [comment]
@gillg Some good news, and some bad news!

Bad news: neither OpenBao nor Vault allow you to accomplish this step:

> Using PKI engine instead of key/value, you can generate a CA cert and private key directly on vault (not locally with openSSL) without never having access to the private key.

The PKI engine is not a _usage_ engine (i.e., it does not hold certificates on behalf of callers), it is a [Certificate Authority](https://datatracker.ietf.org/doc/html/rfc5280#section-3) and thus simply creates certificates.

The [`issue` endpoints](https://openbao.org/api-docs/secret/pki/#generate-certificate-and-key) return and do not store the key material. It is up to the caller to do so and would merely replace the OpenSSL step.

---

That said, the good news is that both OpenBao and Vault have a solution to this: Transit. This is a Key Management System (KMS) type solution with support for PKI Cert association: https://openbao.org/api-docs/secret/transit/#sign-csr + https://openbao.org/api-docs/secret/transit/#set-certificate-chain

This will let you generate key material within OpenBao, create a CSR for some to sign, import the final certificate, and use the corresponding key material whenever desired. K/V thus isn't necessary. 

I had mentioned it here: https://github.com/notaryproject/.github/issues/26#issuecomment-1536565739  and @Gabrielopesantos subsequently implemented it in Vault and ported the feature to OpenBao. 

I think a PR in this direction would definitely be welcome!

---

The interesting question is what our version compatibility policy is. AFAIK we have not yet done a release so we could do things correctly now, given the feature exists in both places. But we might also run afoul of our external version compatibility promises (e.g., to integrate with all supported releases rather than just the latest of 3rd party software). I think Vault 1.15 is getting old (based on https://endoflife.date/hashicorp-vault it is no longer supported), so we could perhaps be fine with this.

Thoughts, @FeynmanZhou ?

**Comment by gillg** [comment]
Thanks for your answer !

> The PKI engine is not a _usage_ engine (i.e., it does not hold certificates on behalf of callers), it is a [Certificate Authority](https://datatracker.ietf.org/doc/html/rfc5280#section-3) and thus simply creates certificates.

I spoke with an Hashicorp Vault engineer some days ago and he said me the opposite... 🤔 
I will double check soon

But anyway, using transit engine is better than nothing. More tricky for a novice to understand, but the result seems the same. What could be missing compared to a real PKI management is the CLR and related things probably.
I just think https://developer.hashicorp.com/vault/api-docs/secret/transit#sign-data matches more to sign a random digest (OCI thumbprint). 
But indeed after some oneoff steps:
1. transit create key (protected private key)
2. sign CSR 
3. set certificate chain

Would require some experimentations, but sounds work on the paper.

---

Also, I was saying 
>  Limitation: you can't use a cert provided by a public root CA as you can't import the private key or make a CSR.

But I think I'm wrong we can import a private key

**Comment by gillg** [comment]
Side note, something what seems not possible on Vault / OpenBao and what I find "perfect" in Azure keyvault is the fact you can get a publicly certified EV cert, following all best practices to be FIPS 140-2 level 2 (required to get an EV cert), and sign anything, even with Notation.

The idea on Azure is:
* Create self signed cert with some basic details and enable the option "HSM private key" (unexportable)
* Generate a CSR (native feature on Azure certs)
* Provide your CSR to your certificate vendor (digicert, entrust, globalsign....)
* Get your public CRT back signed by the vendor after some days
* Import the CRT on Azure to replace the self-signed x509

=> Then you can use the Azure API "sign" (to sign any arbitrary digest) and "getCert" to get the x509 and embded it in your artifacts.

It's the only provider I've found doing that if you don't consider a dedicated HSM (like AWS CloudHSM).
* Azure cert ~$1/month
* CloudHSM ~$1000/month except if you shut it down / start it when needed.

**Comment by cipherboy** [comment]
@gillg Yes, this flow with Azure KeyVault is possible with OpenBao's Transit engine:

1. [Create key in Transit](https://openbao.org/api-docs/secret/transit/#create-key) with `exportable=false`
2. Locally, create a template CSR; provide it to [Transit's sign CSR endpoint](https://openbao.org/api-docs/secret/transit/#sign-csr)
3. Provide signed CSR to your public certificate vendor 
4. Get the signed certificate back.
5. Import the certificate+optionally chain to Transit [using the set-chain endpoint](https://openbao.org/api-docs/secret/transit/#set-certificate-chain). 


The only difference is at step 1, you aren't shipping out a self-signed X.509 certificate, and step 2's interface isn't "any parameter on a CSR" it is "take this existing CSR as a template, swap key material, and sign it with this Transit key".

**Comment by gillg** [comment]
Ah ok ! So it was already exactly the previous scenario you described 😅  
Sorry for the misunderstanding it was less obvious on the paper. The sign CSR endpoint was confusing me because I was thinking the goal was to sign an external CSR to work as a CA.

**Comment by cipherboy** [comment]
@gillg No worries. If you have wording changes for the OpenBao docs, let me know!

**Comment by gillg** [comment]
Purely about wording and clarification, firstable I would try to make work this feature outside of transit engine. 
I know the concept is common... But in cryptography there is no tons of "simple" RFCs in my opinion. If we consider all the non deprecated PKCS#X and their underlying sym/asyn/hash algos + ASN1 + modern envelopes like JWT we have almost all.
But in vault we have multiple entry points for similar underlying concepts.
Why not adding Apis in PKI for that purpose?
Transit is expected to be dedicated for secrets exchange, we don't assume using it for software signature and attestation.

But if you want we can discuss that outside of this issue to not pollute it :)
I would be happy to see with you what can be done to simplify scarring concepts for common devs, to make this kind of security more adopted for a more secured software world 😅
Let's encrypt made things usable for anyone, I really think Bao has a role to play.

**Comment by cipherboy** [comment]
@gillg Ahh, I think the confusion stems from the name "transit" -- Transit is just a branded name for a KMS solution, like Azure KeyVault or GCP CloudKMS, &c.

Transit isn't a secrets storage interface, it is a general purpose cryptographic API. 

K/V (KVv1, KVv2) is what we call our static secret store.
